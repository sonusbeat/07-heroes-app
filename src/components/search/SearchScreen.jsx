import { useLocation } from 'react-router';
import queryString from "query-string";

import heroes from '../../data/heroes';
import HeroCard from '../heroes/HeroCard';
import useForm from '../../hooks/useForm';

const SearchScreen = ({ history }) => {

  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  // const heroesFiltered = heroes.filter(hero => hero.id === 1);
  const heroesFiltered = heroes;

  const [ formValues, handleInputChange, reset ] = useForm({ searchText: q });

  const { searchText } = formValues;

  const HandleSearch = (event) => {
    event.preventDefault();
    history.push(`?q=${searchText}`);
    reset();
  };

  return (
    <section className="animate__animated animate__fadeIn">

      <h1>Search Hero</h1>
      <hr />

      <div className="row">

        <div className="col-5">
          <h2>Search Form</h2>
          <hr/>

          <form onSubmit={ HandleSearch }>
            <div className="form-group">
              <input
                name="searchText"
                type="text"
                className="form-control mb-4"
                placeholder="Find your hero ..."
                onChange={ handleInputChange }
                value={ searchText }
                autoComplete="off"
              />
              <button
                onClick={ HandleSearch }
                className="btn btn-primary btn-block"
              >Search</button>
            </div>
          </form>
        </div>{/* .col */}

        <div className="col-7">
          <h2>Results</h2>
          <hr />
          {
            heroesFiltered.map(hero => (
              <HeroCard key={ hero.id } { ...hero } />
            ))
          }
        </div>{/* .col */}

      </div>{/* .row */}

    </section>
  );
}

export default SearchScreen;