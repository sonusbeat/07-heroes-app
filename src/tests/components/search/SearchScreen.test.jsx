import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Route } from 'react-router-dom';

import SearchScreen from '../../../components/search/SearchScreen';

describe('Pruebas en el componente <Search />', () => {

  test('Deberia mostrarse correctamente con valores por defecto', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route to="search" component={ SearchScreen } />
      </MemoryRouter>
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find(".alert-info").text().trim() ).toBe( "Search a Hero" );

  });

  test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
    const query = "batman";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Route to="search" component={ SearchScreen } />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe( query );
    expect( wrapper ).toMatchSnapshot();

  });

  test('Debe de mostrar la alerta que si no encontró el hero', () => {
    const query = "batasman";
    const alertMessage = `There is no a hero with the following query: "${query}"`;

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Route to="search" component={ SearchScreen } />
      </MemoryRouter>
    );

    expect( wrapper.find(".alert-danger").exists() ).toBe( true );
    expect( wrapper.find(".alert-danger").text().trim() ).toBe( alertMessage );
    expect( wrapper ).toMatchSnapshot();

  });

  test('Debe de llamar el push del history', () => {
    const history = {
      push: jest.fn(),
    };

    const query = "ironman";

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Route
          to="search"
          component={ () => <SearchScreen history={ history } /> }
        />
      </MemoryRouter>
    );

    // Simulación cambio en la caja de texto
    wrapper.find("input").simulate('change', {
      name: "searchText",
      value: query
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault(){}
    });

    expect( history.push ).toHaveBeenCalledWith(`?q=${query}`);

  });


});
