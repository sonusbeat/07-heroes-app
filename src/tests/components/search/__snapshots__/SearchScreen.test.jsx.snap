// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Search /> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<section
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h2>
        Search Form
      </h2>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control mb-4"
            name="searchText"
            onChange={[Function]}
            placeholder="Find your hero ..."
            type="text"
            value="batman"
          />
          <button
            className="btn btn-primary btn-block"
            onClick={[Function]}
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h2>
        Results
      </h2>
      <hr />
      <section
        className="card md-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h3
                className="card-title"
              >
                Batman
              </h3>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Más ...
              </a>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</section>
`;

exports[`Pruebas en el componente <Search /> Debe de mostrar la alerta que si no encontró el hero 1`] = `
<section
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h2>
        Search Form
      </h2>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control mb-4"
            name="searchText"
            onChange={[Function]}
            placeholder="Find your hero ..."
            type="text"
            value="batasman"
          />
          <button
            className="btn btn-primary btn-block"
            onClick={[Function]}
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with the following query: "
        batasman
        "
      </div>
    </div>
  </div>
</section>
`;

exports[`Pruebas en el componente <Search /> Deberia mostrarse correctamente con valores por defecto 1`] = `
<section
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h2>
        Search Form
      </h2>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control mb-4"
            name="searchText"
            onChange={[Function]}
            placeholder="Find your hero ..."
            type="text"
            value=""
          />
          <button
            className="btn btn-primary btn-block"
            onClick={[Function]}
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</section>
`;
